edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/authenticators/DeviceScopeTokenAuthenticator.cs

```diff
@@ -31,62 +31,158 @@ public class DeviceScopeTokenAuthenticator : DeviceScopeAuthenticator<ITokenCred
 
-        protected override bool ValidateWithServiceIdentity(ServiceIdentity serviceIdentity, ITokenCredentials credentials) =>
+        protected override bool ValidateWithServiceIdentity(ServiceIdentity serviceIdentity, ITokenCredentials credentials, Option<string> authChain) =>
             this.TryGetSharedAccessSignature(credentials.Token, credentials.Identity, out SharedAccessSignature sharedAccessSignature)
-                ? this.ValidateCredentials(sharedAccessSignature, serviceIdentity, credentials.Identity)
+                ? this.ValidateCredentials(sharedAccessSignature, serviceIdentity, credentials.Identity, authChain)
                 : false;
 
-        internal bool ValidateAudience(string audience, IIdentity identity)
-        {
+        internal (string hostName, string deviceId, Option<string> moduleId) ParseAudience(string audience)
+        {
             Preconditions.CheckNonWhiteSpace(audience, nameof(audience));
             audience = WebUtility.UrlDecode(audience.Trim());
+
             // The audience should be in one of the following formats -
             // {HostName}/devices/{deviceId}/modules/{moduleId}
             // {HostName}/devices/{deviceId}
             string[] parts = audience.Split('/');
             string hostName;
+            string deviceId;
+            Option<string> moduleId = Option.None<string>();
             if (parts.Length == 3)
             {
                 hostName = parts[0];
-                string deviceId = parts[2];
-                if (identity is IDeviceIdentity deviceIdentity && deviceIdentity.DeviceId != deviceId)
+                deviceId = parts[2];
+            }
+            else if (parts.Length == 5)
+            {
+                hostName = parts[0];
+                deviceId = parts[2];
+                moduleId = Option.Some(parts[4]);
+            }
+            else
+            {
+                throw new ArgumentException($"Invalid audience: {audience}");
+            }
+
+            return (hostName, deviceId, moduleId);
+        }
+
+        internal bool ValidateAuthChain(string actorDeviceId, Option<string> actorModuleId, string targetId, string authChain)
+        {
+            var authChainIds = authChain.Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
+
+            // The first element of the authChain should be the target identity
+            if (authChainIds[0] != targetId)
+            {
+                Events.InvalidAuthChain(targetId, authChain);
+                return false;
+            }
+
+            // The actor identity should be in the authChain
+            string actorId = actorDeviceId + actorModuleId.Map(id => "/" + id).GetOrElse(string.Empty);
+            bool targetAuthChainHasActor = false;
+            foreach (string id in authChainIds)
+            {
+                if (id == actorDeviceId)
```

Comment:
>actorDeviceId [](start = 26, length = 13)

this should be actorId

Reply:
actually actorDeviceId is correct here, the authchain won't have the acting module ($edgeHub), only the deviceId.  The actorId local var is extraneous.

---
In reply to: [452482339](https://github.com/Azure/iotedge/pull/3190#discussion_r452482339) [](ancestors = 452482339)